aten/src/ATen/native/cpu/Activation.cpp

```diff
@@ -1,31 +1,81 @@
 #include <ATen/native/Activation.h>
 
 #include <ATen/ATen.h>
+#include <ATen/Config.h>
 #include <ATen/cpu/vec256/vec256.h>
 #include <ATen/native/TensorIterator.h>
 #include <ATen/native/cpu/Loops.h>
 
-namespace at { namespace native {
+#if AT_MKL_ENABLED()
+#include <mkl.h>
+#endif // AT_MKL_ENABLED()
+
+namespace at {
+namespace native {
+
 namespace {
 
-static void threshold_kernel(TensorIterator& iter, Scalar threshold_scalar, Scalar value_scalar) {
+static void threshold_kernel(
+    TensorIterator& iter,
+    Scalar threshold_scalar,
+    Scalar value_scalar) {
   AT_DISPATCH_ALL_TYPES(iter.dtype(), "threshold_cpu", [&] {
     using Vec = Vec256<scalar_t>;
     scalar_t threshold = threshold_scalar.to<scalar_t>();
     scalar_t value = value_scalar.to<scalar_t>();
     binary_kernel_vec(
-      iter,
-      [&](scalar_t x, scalar_t other) -> scalar_t {
-        return x <= threshold ? value : other;
-      },
-      [&](Vec x, Vec other) -> Vec {
-        return Vec::blendv(other, Vec(value), x <= Vec(threshold));
-      });
+        iter,
+        [&](scalar_t x, scalar_t other) -> scalar_t {
+          return x <= threshold ? value : other;
+        },
+        [&](Vec x, Vec other) -> Vec {
+          return Vec::blendv(other, Vec(value), x <= Vec(threshold));
+        });
   });
 }
 
-} // anonymous namespace
+#if AT_MKL_ENABLED()
+
+template <typename T>
+void GeluKernelMKL(const Tensor& X, Tensor* Y);
+
+#define DELEGATE_GELU_KERNEL_MKL(T, CdfNormFunc, MulFunc) \
+  template <>                                             \
+  void GeluKernelMKL<T>(const Tensor& X, Tensor* Y) {     \
+    const int64_t N = X.numel();                          \
+    const T* X_data = X.data<T>();                        \
+    T* Y_data = Y->data<T>();                             \
+    CdfNormFunc(N, X_data, Y_data);                       \
+    MulFunc(N, X_data, Y_data, Y_data);                   \
+  }
+DELEGATE_GELU_KERNEL_MKL(float, vsCdfNorm, vsMul)
+DELEGATE_GELU_KERNEL_MKL(double, vdCdfNorm, vdMul)
+#undef DELEGATE_GELU_KERNEL_MKL
+
+#else // AT_MKL_ENABLED()
+
+template <typename T>
+void GeluKernelMKL(const Tensor& X, Tensor* Y) {
+  AT_ASSERTM(false, "ATen not compiled with MKL");
+}
+
+#endif // AT_MKL_ENABLED()
+
+void GeluKernelImpl(const Tensor& X, Tensor* Y) {
+  if (at::hasMKL()) {
+    AT_DISPATCH_FLOATING_TYPES(X.scalar_type(), "GeluKernelImpl", [&]() {
+      GeluKernelMKL<scalar_t>(X, Y);
```

Comment:
Don't we need to check that the tensors are contiguous before dispatching to the MKL-optimized codepath?

Reply:
This needs a comment; I would expect something like:
1) describe if MKL supports non-contiguous inputs or not.
2a) if it doesn't, when is it worth it to make the tensor contiguous to do the op?  Does it ever or should I just check contiguity in the MKL pass?
2b) if it does, why don't I just pass in the tensor?

Reply:
I will add the cuda impl in the next PR. Thanks for the advice.